ReactWindows/ReactNative.Shared/UIManager/DynamicAutomationPeer.cs

```diff
@@ -0,0 +1,62 @@
+using System.Linq;
+#if WINDOWS_UWP
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Automation.Peers;
+using Windows.UI.Xaml.Automation.Provider;
+#else
+using System.Windows;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+#endif
+
+namespace ReactNative.UIManager
+{
+    /// <summary>
+    /// Automation peer that allows to dynamically change behavior.
+    /// </summary>
+    /// <typeparam name="T">Type of owner of the automation peer.</typeparam>
+    public sealed class DynamicAutomationPeer<T> : FrameworkElementAutomationPeer, IInvokeProvider
+        where T : FrameworkElement, IAccessible
+    {
+        /// <summary>
+        /// Hides base UIElement Owner to provide stronger-typed T Owner.
+        /// </summary>
+        private new T Owner => (T)base.Owner;
+
+        /// <inheritdoc />
+        public DynamicAutomationPeer(T owner)
+            : base(owner)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            if (Owner.AccessibilityTraits?.Contains(AccessibilityTrait.Button) == true)
+            {
+                return AutomationControlType.Button;
+            }
+            return AutomationControlType.Group;
+        }
+
+        /// <inheritdoc />
+        protected override object GetPatternCore(PatternInterface patternInterface)
+        {
+            if (patternInterface == PatternInterface.Invoke
+                && Owner.AccessibilityTraits?.Contains(AccessibilityTrait.Button) == true)
```

Comment:
As i said elsewhere, making AccessiblityTraits non-null but empty may simplify the check

Reply:
I'll copy here my answer to the comment you mention to simplify discussion:

using empty array will actually complicate code since then you would need to check for both null and empty array. First, these two cases have same semantics here, so having them as two different conditions is unneeded complexity. Second, empty array takes memory, and you cannot even add to it in this model, so it's totally unused memory.

---
In reply to: [168885140](https://github.com/gmangov/react-native-windows/pull/1#discussion_r168885140) [](ancestors = 168885140)